### ISSUE TITLE
graph_transforms/transform_graph failed with shared backbone and multiple input node

### ISSUE THREAD

**Issue by hwenjun18** [tensorflow/tensorflow#31877]
**I try to optimize my model with tf transform_graph tool, and failed**

bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=inference.pb \
--out_graph=inference_opt.pb \
--inputs='input_image_1:0,input_image2:0,input_image_3:0’ \
--outputs='output_node_1:0,output_node_2:0’ \
--transforms='
  remove_nodes(op=Identity, op=CheckNumerics)
  fold_batch_norms
  fold_old_batch_norms
  sort_by_execution_order'

**Then long log print occur and never stop, something like this:**

2019-08-22 10:56:11.108570: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/beta/read/_135__cf__135 to be preserved.
2019-08-22 10:56:11.108594: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/gamma/read/_136__cf__136 to be preserved.
2019-08-22 10:56:11.108604: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/moving_mean/read/_137__cf__137 to be preserved.
2019-08-22 10:56:11.108614: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/moving_variance/read/_138__cf__138 to be preserved.
2019-08-22 10:56:11.108623: W tensorflow/tools/graph_transforms/transform_utils.cc:448] Generator function didn't preserve needed nodes, copying old replacements back in instead.
2019-08-22 10:56:11.120159: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/beta/read/_140__cf__140 to be preserved.
2019-08-22 10:56:11.120181: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/gamma/read/_141__cf__141 to be preserved.
2019-08-22 10:56:11.120192: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/moving_mean/read/_142__cf__142 to be preserved.
2019-08-22 10:56:11.120201: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/moving_variance/read/_143__cf__143 to be preserved.
2019-08-22 10:56:11.120211: W tensorflow/tools/graph_transforms/transform_utils.cc:448] Generator function didn't preserve needed nodes, copying old replacements back in instead.
2019-08-22 10:56:11.138263: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/Conv_1/BatchNorm/beta/read/_10__cf__10 to be preserved.
2019-08-22 10:56:11.138286: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/Conv_1/BatchNorm/gamma/read/_11__cf__11 to be preserved.
2019-08-22 10:56:11.138294: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/Conv_1/BatchNorm/moving_mean/

**And after I remove  transform option  **remove_nodes(op=Identity, op=CheckNumerics)**,
It seems succeed, but when I open optimized pb fie, bn node is still there, which not be fold.**
I test other classification model,  transform_graph just woks well. But in my model, it just can not work. **By the way, my model has multiple input and shares backbone mobilenet_v2**.

bazel-bin/tensorflow/tools/graph_transforms/transform_graph \
--in_graph=inference.pb \
--out_graph=inference_opt.pb \
--inputs='input_image_1:0,input_image2:0,input_image_3:0’ \
--outputs='output_node_1:0,output_node_2:0’ \
--transforms='
  fold_batch_norms
  fold_old_batch_norms
  sort_by_execution_order'

**Comment by ravikyram** [comment]
Please provide details about what platform you are using (operating system, architecture). Also include your TensorFlow version. Also, did you compile from source or install a binary?

Please fill the issue template. Could you update them if they are relevant in your case, or leave them as N/A? . If you are unclear what to include see the issue template displayed in [the Github new issue template](https://github.com/tensorflow/tensorflow/issues/new/choose).

We ask for this in the issue submission template, because it is really difficult to help without that information. Thanks!

**Comment by hwenjun18** [comment]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Mac OS 10.14
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.14.0
- Python version: 3.6.5
- Bazel version (if compiling from source): 0.26.0
- GCC/Compiler version (if compiling from source): Apple LLVM version 10.0.0 (clang-1000.11.45.5)
- CUDA/cuDNN version:  No GPU
- GPU model and memory: No GPU


**Describe the current behavior**

I try to optimize my model with tf transform_graph tool, and failed

bazel-bin/tensorflow/tools/graph_transforms/transform_graph 
--in_graph=inference.pb 
--out_graph=inference_opt.pb 
--inputs='input_image_1:0,input_image2:0,input_image_3:0’ 
--outputs='output_node_1:0,output_node_2:0’ 
--transforms='
remove_nodes(op=Identity, op=CheckNumerics)
fold_batch_norms
fold_old_batch_norms
sort_by_execution_order'

Then long log print occur and never stop, something like this:

2019-08-22 10:56:11.108570: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/beta/read/_135__cf__135 to be preserved.
2019-08-22 10:56:11.108594: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/gamma/read/_136__cf__136 to be preserved.
2019-08-22 10:56:11.108604: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/moving_mean/read/_137__cf__137 to be preserved.
2019-08-22 10:56:11.108614: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/expand/BatchNorm/moving_variance/read/_138__cf__138 to be preserved.
2019-08-22 10:56:11.108623: W tensorflow/tools/graph_transforms/transform_utils.cc:448] Generator function didn't preserve needed nodes, copying old replacements back in instead.
2019-08-22 10:56:11.120159: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/beta/read/_140__cf__140 to be preserved.
2019-08-22 10:56:11.120181: W tensorflow/tools/graph_transforms/transform_utils.cc:441] Expected part_cnn/feature_extract/MobilenetV2/expanded_conv_16/project/BatchNorm/gamma/read/_141__cf__141 to be preserved.

**Describe the expected behavior**

Bn should be fold into conv

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by jvishnuvardhan** [comment]
@hwenjun18 Could you share minimum standalone code to reproduce the issue? Did you check graph transform with other tensorflow models? Thanks!

**Comment by hwenjun18** [comment]
> @hwenjun18 Could you share minimum standalone code to reproduce the issue? Did you check graph transform with other tensorflow models? Thanks!

I have checked graph transform with other tensorflow models, including mobilenet_v1, mobilenet_v2 in slim, but with multiple input and a shared mobilenet_v2, it fails

**Comment by hajuho** [comment]
@hwenjun18 Could you add 'fold_constants(ignore_errors=ture)' to the transforms list as fold_constants is prerequisite of fold_batch_norms?
https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms#fold_batch_norms
